src/c4/ifpstream.cc

```diff
@@ -0,0 +1,71 @@
+//----------------------------------*-C++-*-----------------------------------//
+/*!
+ * \file   c4/ifpstream.cc
+ * \author Mathew Cleveland
+ * \date   Mon Jun 25 11:36:43 MDT 2018
+ * \brief  Define methods of class ifpstream
+ * \note   Copyright (C) 2018-2020 Triad National Security, LLC.
+ *         All rights reserved. */
+//----------------------------------------------------------------------------//
+
+#include "ifpstream.hh"
+#include "C4_Functions.hh"
+
+namespace rtt_c4 {
+using namespace std;
+
+//----------------------------------------------------------------------------//
+/*! Create an ifpstream to for a parallel file read.
+ *
+ * Create an ifpstream that reads and brocasts input file streams to all ranks
+ *
+ * \param[in] filename Name of the file to which is being read
+ * \param[in] mode File read mode (ascii/binary)-- defaults to ascii
+ */
+ifpstream::ifpstream(std::string const &filename,
+                     ios_base::openmode const mode) {
+  mode_ = mode;
+  if (rtt_c4::node() == 0) {
+    in_.open(filename, mode);
+  }
+}
+
+//----------------------------------------------------------------------------//
+/*! Fill parallel buffers with data from the input file
+ *
+ * \param[in] local process buffer size
+ */
+void ifpstream::fill_buffers(unsigned const buffer_size) {
+  // initialize N process vector of buffer sizes
+  std::vector<unsigned> buffer_sizes(rtt_c4::nodes(), 0);
+  // fill in local buffer size
+  buffer_sizes[rtt_c4::node()] = buffer_size;
+  // reduce buffer_size to all ranks
+  rtt_c4::global_sum(&buffer_sizes[0], rtt_c4::nodes());
+
+
+  if (rtt_c4::node() == 0) {
+    // populate node zero buffer
+    std::string local_string(buffer_size, '\0');
+    in_.read(&local_string[0], buffer_size);
+    str(local_string);
```

Comment:
This is interesting! I had never seen this version of the str() function before, and had to look it up. So this line is just setting the content of the ifpstream object's stream to local_string? Which you can do because you're inheriting from stringstream? (Sorry if these are dumb questions!)

Reply:
Your exactly right. This is just a fast way to reset the buffer of a stringstream.